internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala

```diff
@@ -533,22 +533,31 @@ case class ProjectStructure(
         val earlyDeps: Future[Seq[Path]] = Future.traverse(dependsOnRef) { dep =>
           dep.earlyArtifact(i).map(success => if (success) dep.earlyOutput else dep.output)
         }
-        val futureAnalysis = earlyDeps.map { internalCp =>
-          doCompile(i, notifyEarlyOutput, internalCp, pipelinedLookupAnalysis)
+        val futureScalaAnalysis = earlyDeps.map { internalCp =>
+          doCompile(i, notifyEarlyOutput, internalCp, pipelinedLookupAnalysis, false)
         }
+        val wholeDeps = Future.traverse(dependsOnRef) { dep =>
+          dep.compile(i).map(_ => dep.output)
+        }
+        def futureJavaAnalysis(projectDependencies: Seq[Path], prev: Analysis): Future[Analysis] =
+          Future {
+            doCompile(i, notifyEarlyOutput, projectDependencies, pipelinedLookupAnalysis, true)
+          }
+
         // wait for the full compilation from the dependencies
         // during pipelining, downstream compilation may complete before the upstream
         // to avoid deletion of directories etc, we need to wait for the upstream to finish
         val f = for {
-          _ <- Future.traverse(dependsOnRef)(_.compile(i))
-          a <- futureAnalysis
+          pj <- wholeDeps
+          a0 <- futureScalaAnalysis
+          a <- futureJavaAnalysis(pj, a0)
         } yield a
```

Comment:
This is the part that waits for upstream whole compilation and self Scala compilation for scripted tests.